[ALLUXIO-2453] Concurrent rename

calvinjia opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-2453

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12620/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-core-server' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12620/org.alluxio$alluxio-core-server/testReport'>org.alluxio:alluxio-core-server</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12620/org.alluxio$alluxio-core-server/testReport/alluxio.master.file/FileSystemMasterTest/folderConcurrentRename/'><strong>alluxio.master.file.FileSystemMasterTest.folderConcurrentRename</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12621/<h2>Build result: FAILURE</span></h2>[...truncated 1897 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.4.0-SNAPSHOT/alluxio-shell-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/pom.xml to org.alluxio/alluxio-underfs-oss/1.4.0-SNAPSHOT/alluxio-underfs-oss-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-oss/1.4.0-SNAPSHOT/alluxio-underfs-oss-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-oss/1.4.0-SNAPSHOT/alluxio-underfs-oss-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/pom.xml to org.alluxio/alluxio-keyvalue-client/1.4.0-SNAPSHOT/alluxio-keyvalue-client-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.4.0-SNAPSHOT/alluxio-underfs-hdfs-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.4.0-SNAPSHOT/alluxio-underfs-hdfs-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.4.0-SNAPSHOT/alluxio-underfs-hdfs-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.4.0-SNAPSHOT/alluxio-underfs-swift-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-swift/1.4.0-SNAPSHOT/alluxio-underfs-swift-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-swift/1.4.0-SNAPSHOT/alluxio-underfs-swift-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/pom.xml to org.alluxio/alluxio-underfs-s3a/1.4.0-SNAPSHOT/alluxio-underfs-s3a-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/target/alluxio-underfs-s3a-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-s3a/1.4.0-SNAPSHOT/alluxio-underfs-s3a-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/target/alluxio-underfs-s3a-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-s3a/1.4.0-SNAPSHOT/alluxio-underfs-s3a-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.4.0-SNAPSHOT/alluxio-examples-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.4.0-SNAPSHOT/alluxio-core-server-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.4.0-SNAPSHOT/alluxio-keyvalue-common-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.4.0-SNAPSHOT/alluxio-underfs-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.4.0-SNAPSHOT/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.4.0-SNAPSHOT/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.4.0-SNAPSHOT/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.4.0-SNAPSHOT/alluxio-integration-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.4.0-SNAPSHOT/alluxio-keyvalue-1.4.0-SNAPSHOT.pomchannel stopped
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12625/
Test PASSed.

apc999 commented on core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -1226,4 +1232,325 @@ private long createFileWithSingleBlock(AlluxioURI uri) throws Exception {
+
+    for (int i = 0; i < numThreads; i++) {
+      srcs[i] = new AlluxioURI("/file" + i);
+      mFileSystemMaster.createFile(srcs[i], sNestedFileOptions);
```
it seems we are able to rename incomplete file. is this expected?

calvinjia replied:
Currently there is nothing stopping us from doing so.

apc999 commented on core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -1226,4 +1232,325 @@ private long createFileWithSingleBlock(AlluxioURI uri) throws Exception {
+
+    int errors = concurrentRename(srcs, dsts);
+
+    Assert.assertEquals("More than 0 errors: " + errors, 0, errors);
```
expect all dsts exist and no srcs exists any more?

calvinjia replied:
Updated, thanks.

apc999 commented on core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -1226,4 +1232,325 @@ private long createFileWithSingleBlock(AlluxioURI uri) throws Exception {
+    }
+    int errors = concurrentRename(srcs, dsts);
+
+    Assert.assertEquals("More than 0 errors: " + errors, 0, errors);
```
expect all dsts exist and no srcs exists any more?

calvinjia replied:
Updated, thanks.

apc999 commented on core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -1226,4 +1232,325 @@ private long createFileWithSingleBlock(AlluxioURI uri) throws Exception {
+
+    // We should get an error for all but 1 rename
+    Assert.assertEquals(numThreads - 1, errors);
+    // Only one renamed file should exist
```
also expect no src exists?

calvinjia replied:
Updated, thanks.

apc999 commented on core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -1226,4 +1232,325 @@ private long createFileWithSingleBlock(AlluxioURI uri) throws Exception {
+    List<FileInfo> dirChildren =
+        mFileSystemMaster.listStatus(new AlluxioURI(existingDirs.get(0).getPath()),
+            ListStatusOptions.defaults());
+    Assert.assertEquals(1, dirChildren.size());
```
expect src dir not exist?

calvinjia replied:
Updated, thanks.

apc999 commented on core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -1226,4 +1232,325 @@ private long createFileWithSingleBlock(AlluxioURI uri) throws Exception {
+        mFileSystemMaster.listStatus(dir2, ListStatusOptions.defaults());
+
+    Assert.assertEquals(0, dir1Files.size());
+    Assert.assertEquals(numThreads, dir2Files.size());
```
not only the file size, but can we also confirm the file names are correct (e.g., they really become `rename{1,2,3...}`?

calvinjia replied:
Updated, thanks.

jsimsa commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -486,18 +487,24 @@ private StartupConsistencyChecker(Long fileId) {
+              childInode.lockReadAndCheckParent(parentInode);
+            } catch (InvalidPathException e) {
+              // This should be safe, continue.
+              LOG.debug("A file scheduled for consistency check was moved before the check.");
```
do we release the lock on the inode if the exception is thrown?

calvinjia replied:
The startup check does not block until complete, so the user can possibly do other write operations while the startup check is running. We cannot log the path because we do not have a lock on the inode (childInode.getName is not guaranteed to be accurate). We could log the parent directory, but I don't think this is would be helpful.

calvinjia replied:
If an exception is thrown, the lock will be unlocked in `lockReadAndCheckParent`.

jsimsa replied:
ack

jsimsa replied:
would it make sense to log e.message() instead of assuming the root case of this exception?

calvinjia replied:
Good point, updated.

jsimsa commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1509,13 +1516,20 @@ private boolean isFullyInMemory(InodeFile inode) {
+    try {
+      root.lockReadAndCheckParent(null);
+    } catch (InvalidPathException e) {
+      // This should never happen.
```
log the error?

calvinjia replied:
Removed due to Bin's suggestion.

jsimsa commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1830,65 +1851,110 @@ private long renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
+      // On failure, revert changes and throw exception.
+      srcInode.setName(srcName);
+      srcInode.setParentId(dstParentInode.getId());
+      if (!dstParentInode.removeChild(dstName)) {
```
could this ever happen?

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -192,8 +193,22 @@ public T setGroup(String group) {
+
+  /**
+   * @param lastModificationTimeMs the last modification time to use
+   * @param override if true, sets the value regardless of the previous last modified time,
```
why do we introduce this flag?

calvinjia replied:
When replaying from the journal, we create inodes which have default `lastModificationTimeMs` to be `System.currentTimeMillis` and then set the modification time to the correct time. If you prefer we could keep the old `setLastModificationTimeMs` and introduce a new method called `updateLastModificationTimeMs` which has the logic of only taking newer times.

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,24 +292,93 @@ public T setOwner(String owner) {
+   * and not by path or parent.
    */
   public void lockWrite() {
+    mLock.readLock().lock();
```
writeLock?

jsimsa replied:
did our unit / integration tests not catch this? if not, please add a test that would catch this next time

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/InodeLockList.java:
```diff
@@ -33,9 +35,27 @@
    *
    * @param inode the inode to lock
+   * @param parent the expected parent inode
+   * @throws InvalidPathException if the inode is no long consistent with the caller's expectations
```
no longer

jsimsa replied:
it might make sense to be more explicit about what the caller expectations are between the different versions of the overloaded method

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/InodeLockList.java:
```diff
@@ -33,9 +35,27 @@
+   * @param inode the inode to lock
+   * @param parent the expected parent inode
+   * @param name the expected name of the inode to be locked
+   * @throws InvalidPathException if the inode is no long consistent with the caller's expectations
```
no longer

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/InodeLockList.java:
```diff
@@ -60,9 +80,27 @@ public synchronized void unlockLast() {
    * @param inode the inode to lock
+   * @param parent the expected parent inode
+   * @param name the expected name of the inode to be locked
+   * @throws InvalidPathException if the inode is no long consistent with the caller's expectations
```
no longer

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/InodeLockList.java:
```diff
@@ -60,9 +80,27 @@ public synchronized void unlockLast() {
    */
-  public synchronized void lockWrite(Inode<?> inode) {
-    inode.lockWrite();
+  public synchronized void lockWriteAndCheckParent(Inode<?> inode, Inode parent)
```
nit: swap the order of the two overloaded lockWrite methods

calvinjia replied:
I previously wanted to respect alphabetical order for the sake of consistency, but I agree it is better to put the method with fewer arguments first despite not being an overloaded method.

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,24 +292,93 @@ public T setOwner(String owner) {
+   * @param name the expected name of the inode to be locked
+   * @throws InvalidPathException if the parent and/or name is not as expected
+   */
+  public void lockWriteAndCheckFullPath(Inode parent, String name)
```
does this really check the full path? what if the parent of the parent is renamed? the full path has changed but this method will not throw an exception

calvinjia replied:
See reply above.

jsimsa replied:
ack

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,24 +292,93 @@ public T setOwner(String owner) {
+   * @param name the expected name of the inode to be locked
+   * @throws InvalidPathException if the parent and/or name is not as expected
+   */
+  public void lockReadAndCheckFullPath(Inode parent, String name) throws InvalidPathException {
```
does this really check the full path? what if the parent of the parent is renamed? the full path has changed but this method will not throw an exception

jsimsa replied:
ack

jsimsa replied:
actually, this call on its own does not guarantee that we are holding a lock on the parent. If that is an assumption to correct operation of this method, then it should be documented.

calvinjia replied:
Updated, thanks.

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/InodeTree.java:
```diff
@@ -946,9 +961,11 @@ private TraversalResult traverseToInode(String[] pathComponents, LockMode lockMo
               == LockMode.READ) {
-            lockList.lockRead(mRoot);
+            // Do not need to check parent for root.
+            lockList.lockReadAndCheckParent(mRoot, null);
```
it might make sense to define SKIP_ROOT_PARENT_CHECK = null someplace and use it instead of the comment

jsimsa commented on core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -1226,4 +1232,325 @@ private long createFileWithSingleBlock(AlluxioURI uri) throws Exception {
+   *
+   * @param src list of source paths
+   * @param dst list of destination paths
+   * @return the duration of the rename and how many errors occurred
```
no duration is returned

calvinjia replied:
Updated, thanks.

jsimsa commented on tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java:
```diff
@@ -399,7 +399,7 @@ public void getCapacityBytes() {
   public void lastModificationTimeCompleteFile() throws Exception {
     long fileId = mFsMaster.createFile(new AlluxioURI("/testFile"), CreateFileOptions.defaults());
-    long opTimeMs = TEST_CURRENT_TIME;
+    long opTimeMs = System.currentTimeMillis() + TEST_TIME_OFFSET_MS;
```
why are we using system time? this could lead to test brittleness

calvinjia replied:
I could use max long value here instead if you prefer. The change is because we need to have an operation time which is later than the creation time since we no longer update the last mod time if it is stale during rename (due to allowing for concurrent renames).

jsimsa replied:
I would prefer to have a race on last modification time instead of the added complexity

jsimsa replied:
I don't feel strongly about that though

jsimsa replied:
let's use max long in case you want to keep the new semantics

calvinjia replied:
I will use long max value. The race conditions could lead to some nontrivial last modification time skew because the modification time used for renames is recorded before any ufs operations.

calvinjia commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1509,13 +1516,20 @@ private boolean isFullyInMemory(InodeFile inode) {
+    Inode root = mInodeTree.getRoot();
+    // Root has no real parent, use null.
+    try {
+      root.lockReadAndCheckParent(null);
```
Good point, updated.

calvinjia commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1509,13 +1516,20 @@ private boolean isFullyInMemory(InodeFile inode) {
 
   private void getInMemoryFilesInternal(Inode<?> inode, AlluxioURI uri,
       List<AlluxioURI> inMemoryFiles) {
-    inode.lockRead();
```
Decided to go with option 2 in order to avoid locking and unlocking inodes across function calls.

calvinjia commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1830,65 +1851,110 @@ private long renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
     InodeDirectory srcParentInode = srcInodePath.getParentInodeDirectory();
+    InodeDirectory dstParentInode = dstInodePath.getParentInodeDirectory();
+    String srcName = srcInode.getName();
+    String dstName = dstPath.getName();
```
The difference is because the destination inode does not exist yet.

calvinjia commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1830,65 +1851,110 @@ private long renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
-    srcParentInode.setLastModificationTimeMs(opTimeMs);
-    srcInode.setParentId(dstParentInode.getId());
+
+    // 4. Remove the source inode (reverting the name) from the source parent.
```
Updated

calvinjia commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1830,65 +1851,110 @@ private long renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
+        LOG.error("Failed to revert rename changes. Alluxio metadata may be inconsistent.");
+      }
+    }
     srcInode.setName(dstPath.getName());
```
Updated, thanks.

calvinjia commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1830,65 +1851,110 @@ private long renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
+      srcInode.setName(srcName);
+      srcInode.setParentId(dstParentInode.getId());
+      if (!dstParentInode.removeChild(dstName)) {
+        LOG.error("Failed to revert rename changes. Alluxio metadata may be inconsistent.");
```
This should not happen, currently in our code we do not consider cases where this call would return false.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12653/
Test PASSed.

calvinjia commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,24 +292,93 @@ public T setOwner(String owner) {
+    lockReadAndCheckParent(parent);
+    if (!mName.equals(name)) {
+      mLock.readLock().unlock();
+      throw new InvalidPathException("Path is no longer valid, possibly due to a concurrent move.");
```
Updated, thanks.

calvinjia commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,24 +292,93 @@ public T setOwner(String owner) {
+    lockWriteAndCheckParent(parent);
+    if (!mName.equals(name)) {
+      mLock.writeLock().unlock();
+      throw new InvalidPathException("Path is no longer valid, possibly due to a concurrent move.");
```
Updated, thanks.

calvinjia commented:
@apc999 @jsimsa Thanks for the reviews, I've address the comments.

jsimsa commented on core/common/src/main/java/alluxio/exception/ExceptionMessage.java:
```diff
@@ -119,6 +119,7 @@
   NOT_MUTABLE_INODE_PATH("Not a MutableLockedInodePath: {0}"),
   PATH_COMPONENTS_INVALID("Parameter pathComponents is {0}"),
   PATH_COMPONENTS_INVALID_START("Path starts with {0}"),
+  PATH_INVALID_CONCURRENT_MOVE("Path is no longer valid, possibly due to a concurrent move."),
```
should we use "rename" consistently?

calvinjia replied:
Updated, thanks.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12662/
Test FAILed.

jsimsa commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -486,18 +487,24 @@ private StartupConsistencyChecker(Long fileId) {
-              childInode.lockRead();
-              try {
+            try {
+              childInode.lockReadAndCheckParent(parentInode);
```
could we return a Closeable from the different lock calls so that try-with-resource idiom can be used for releasing the lock?

calvinjia replied:
One issue I noticed with our current master code is that we do not always abstract the inode lock. Ideally we should always use `LockedInodePath` (which is an `AutoClosable`), but in certain places we directly use the low level API (ie. `inode.lockRead()`) which gets us into these less elegant pieces of code. I suggest we refactor these instances in a separate PR instead of tagging it onto this one, since I think we would need to expand the utility provided in LockedInodePath.

jsimsa replied:
ack

calvinjia replied:
I've created this ticket to track the issue: https://alluxio.atlassian.net/browse/ALLUXIO-2458

jsimsa commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1508,29 +1515,48 @@ private boolean isFullyInMemory(InodeFile inode) {
+      // This inode is a directory.
+      Set<Inode<?>> children = ((InodeDirectory) inode).getChildren();
+      for (Inode<?> child : children) {
+        try {
```
I am not big fan of this "try - catch - try - finally" pattern. If we the lock method returned a closeable, the whole block would look like this:

```
try (InodeLock lock = inode.lockAndCheck();) {
  getInMemoryFilesInternal();
} catch (LockCheckFailed e) {
  continue;
}
```

I also propose to used a specific exception so that the try block in the above expression can include methods that could throw InvalidPathException

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12661/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-core-server' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12661/org.alluxio$alluxio-core-server/testReport'>org.alluxio:alluxio-core-server</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12661/org.alluxio$alluxio-core-server/testReport/alluxio.master.file/FileSystemMasterTest/sameDirConcurrentRename/'><strong>alluxio.master.file.FileSystemMasterTest.sameDirConcurrentRename</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12663/
Test PASSed.

calvinjia commented on core/server/src/main/java/alluxio/master/file/meta/InodeTree.java:
```diff
@@ -946,8 +961,10 @@ private TraversalResult traverseToInode(String[] pathComponents, LockMode lockMo
         if (pathComponents[0].equals("")) {
           if (getLockModeForComponent(0, pathComponents.length, lockMode, lockHints)
               == LockMode.READ) {
+            // Do not need to check parent for root.
```
Sure, the conditions in which to use `lockRead/Write` are documented there so it should not be confusing.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12665/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12664/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12659/<h2>Build result: ABORTED</span></h2>[...truncated 4226 lines...][INFO] Alluxio Under File System - S3 .................... SUCCESS [10.537s][INFO] Alluxio Under File System - S3A ................... SUCCESS [10.544s][INFO] Alluxio Under File System - GCS ................... SUCCESS [9.509s][INFO] Alluxio Core - Client ............................. SUCCESS [55.890s][INFO] Alluxio Core - Server ............................. SUCCESS [1:25.597s][INFO] Alluxio Key Value ................................. SUCCESS [1.576s][INFO] Alluxio Key Value - Common Utilities .............. SUCCESS [14.094s][INFO] Alluxio Key Value - Client ........................ SUCCESS [11.837s][INFO] Alluxio Key Value - Server ........................ SUCCESS [9.256s][INFO] Alluxio Shell ..................................... SUCCESS [11.602s][INFO] Alluxio Examples .................................. SUCCESS [10.677s][INFO] Alluxio MiniCluster ............................... SUCCESS [9.054s][INFO] Alluxio Tests ..................................... FAILURE [1:53:22.252s][INFO] Alluxio Integration ............................... SKIPPED[INFO] Alluxio Integration - Mesos ....................... SKIPPED[INFO] Alluxio Integration - YARN ........................ SKIPPED[INFO] Alluxio Assemblies ................................ SKIPPED[INFO] ------------------------------------------------------------------------[INFO] BUILD FAILURE[INFO] ------------------------------------------------------------------------[INFO] Total time: 1:59:56.568s[INFO] Finished at: Mon Dec 19 15:11:21 PST 2016[INFO] Final Memory: 85M/2323M[INFO] ------------------------------------------------------------------------
Test FAILed.

calvinjia commented:
Jenkins, test this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12669/
Test PASSed.

apc999 commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1830,65 +1857,114 @@ private long renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
+
+    // 5. Set the last modification times for both source and destination parent inodes.
     dstParentInode.setLastModificationTimeMs(opTimeMs);
+    srcParentInode.setLastModificationTimeMs(opTimeMs);
```
since we don't have WRITE lock on dst and src parent inode, I noticed that `setLastModificationTimeMs` now becomes thread-safe which is not documented. Could you update both the comment here (saying this requires thread-safe on parerent inodes) and the java doc on `setLastModificationTimeMs`

calvinjia replied:
Updated, thanks

apc999 commented on core/server/src/main/java/alluxio/master/file/meta/InodeTree.java:
```diff
@@ -565,7 +565,7 @@ public CreatePathResult createPath(LockedInodePath inodePath, CreatePathOptions<
               currentInodeDirectory.getId(), pathComponents[k], missingDirOptions);
       // Lock the newly created inode before subsequent operations, and add it to the lock group.
-      lockList.lockWrite(dir);
+      lockList.lockWriteAndCheckFullPath(dir, currentInodeDirectory, pathComponents[k]);
```
- Is it possible to see `InvalidPathException` here due to concurrent rename? shall we retry if `InvalidPathException` is thrown? It seems now the error is directly thrown back to user.  
- More generally, in some other methods, the same exception is caught and continue. Do we have a guideline when we ignore errors and when not?

apc999 replied:
Is it the case that, when locking an inode during an iteration, we ignore the returned `InvalidPathException`; while for all the other references, we escalate the exception? If yes, we should document it somewhere for future reference.

calvinjia replied:
I've documented this in FileSystemMaster's locking docs portion

apc999 replied:
ack

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12678/<h2>Build result: FAILURE</span></h2>[...truncated 1864 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.4.0-SNAPSHOT/alluxio-shell-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/pom.xml to org.alluxio/alluxio-underfs-oss/1.4.0-SNAPSHOT/alluxio-underfs-oss-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-oss/1.4.0-SNAPSHOT/alluxio-underfs-oss-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-oss/1.4.0-SNAPSHOT/alluxio-underfs-oss-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/pom.xml to org.alluxio/alluxio-keyvalue-client/1.4.0-SNAPSHOT/alluxio-keyvalue-client-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.4.0-SNAPSHOT/alluxio-underfs-hdfs-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.4.0-SNAPSHOT/alluxio-underfs-hdfs-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.4.0-SNAPSHOT/alluxio-underfs-hdfs-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.4.0-SNAPSHOT/alluxio-underfs-swift-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-swift/1.4.0-SNAPSHOT/alluxio-underfs-swift-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-swift/1.4.0-SNAPSHOT/alluxio-underfs-swift-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/pom.xml to org.alluxio/alluxio-underfs-s3a/1.4.0-SNAPSHOT/alluxio-underfs-s3a-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/target/alluxio-underfs-s3a-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-s3a/1.4.0-SNAPSHOT/alluxio-underfs-s3a-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/target/alluxio-underfs-s3a-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-s3a/1.4.0-SNAPSHOT/alluxio-underfs-s3a-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.4.0-SNAPSHOT/alluxio-examples-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.4.0-SNAPSHOT/alluxio-core-server-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.4.0-SNAPSHOT/alluxio-keyvalue-common-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.4.0-SNAPSHOT/alluxio-underfs-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.4.0-SNAPSHOT/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.4.0-SNAPSHOT/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.4.0-SNAPSHOT/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.4.0-SNAPSHOT/alluxio-integration-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.4.0-SNAPSHOT/alluxio-keyvalue-1.4.0-SNAPSHOT.pomchannel stopped
Test FAILed.

calvinjia commented:
@apc999 @jsimsa 
Thanks for the reviews, I've addressed the comments. I will add the more sophisticated concurrent test in a separate PR. (https://alluxio.atlassian.net/browse/ALLUXIO-2459)

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12679/<h2>Build result: FAILURE</span></h2>[...truncated 1863 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.4.0-SNAPSHOT/alluxio-shell-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/pom.xml to org.alluxio/alluxio-underfs-oss/1.4.0-SNAPSHOT/alluxio-underfs-oss-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-oss/1.4.0-SNAPSHOT/alluxio-underfs-oss-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-oss/1.4.0-SNAPSHOT/alluxio-underfs-oss-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/pom.xml to org.alluxio/alluxio-keyvalue-client/1.4.0-SNAPSHOT/alluxio-keyvalue-client-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.4.0-SNAPSHOT/alluxio-underfs-hdfs-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.4.0-SNAPSHOT/alluxio-underfs-hdfs-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.4.0-SNAPSHOT/alluxio-underfs-hdfs-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.4.0-SNAPSHOT/alluxio-underfs-swift-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-swift/1.4.0-SNAPSHOT/alluxio-underfs-swift-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-swift/1.4.0-SNAPSHOT/alluxio-underfs-swift-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/pom.xml to org.alluxio/alluxio-underfs-s3a/1.4.0-SNAPSHOT/alluxio-underfs-s3a-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/target/alluxio-underfs-s3a-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-s3a/1.4.0-SNAPSHOT/alluxio-underfs-s3a-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/target/alluxio-underfs-s3a-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-s3a/1.4.0-SNAPSHOT/alluxio-underfs-s3a-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.4.0-SNAPSHOT/alluxio-examples-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.4.0-SNAPSHOT/alluxio-core-server-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.4.0-SNAPSHOT/alluxio-keyvalue-common-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.4.0-SNAPSHOT/alluxio-underfs-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.4.0-SNAPSHOT/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.4.0-SNAPSHOT/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.4.0-SNAPSHOT/alluxio-underfs-glusterfs-1.4.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.4.0-SNAPSHOT/alluxio-integration-1.4.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.4.0-SNAPSHOT/alluxio-keyvalue-1.4.0-SNAPSHOT.pomchannel stopped
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12680/
Test PASSed.

apc999 commented on core/server/src/test/java/alluxio/master/file/meta/InodeFileTest.java:
```diff
@@ -167,4 +174,214 @@ public void permissionStatus() {
+   * Tests the {@link Inode#lockReadAndCheckFullPath(Inode, String)} method.
+   */
+  @Test
+  public void lockReadAndCheckFullPath() throws Exception {
```
if we merge `inode.lockReadAndCheckFullPath()` to `inodeList.lockReadAndCheckFullPath()`, we will then test this method there, which seems makes more sense as we could assert the lock state on the list

apc999 commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,27 +296,100 @@ public T setOwner(String owner) {
+   * path is consistent with what the caller is expecting. If the state is inconsistent, an
+   * exception will be thrown and the lock will be released.
+   *
+   * This method assumes that the inode path to the parent has been read locked.
```
read -> write

calvinjia replied:
We only need a read lock on the parent.

jsimsa replied:
nit: `NOTE: ...`

apc999 commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,27 +296,100 @@ public T setOwner(String owner) {
+   */
+  public void lockReadAndCheckFullPath(Inode parent, String name) throws InvalidPathException {
+    lockReadAndCheckParent(parent);
+    if (!mName.equals(name)) {
```
I am also confused by the name. I think compared to `lockReadAndCheckParent`, this method can be named after  something like `lockReadAndCheckNameAndParent`, which has less to do with full path.

calvinjia replied:
I called it check full path because with all the implicit conditions, checking the name and parent ensures the entire path is consistent. However I can see it can be confusing and I'm open to changing the name to something more descriptive of the actual operation.

apc999 commented on core/server/src/main/java/alluxio/master/file/meta/InodeTree.java:
```diff
@@ -585,7 +585,7 @@ public CreatePathResult createPath(LockedInodePath inodePath, CreatePathOptions<
     if (lastInode != null) {
       // Lock the last inode before subsequent operations, and add it to the lock group.
-      lockList.lockWrite(lastInode);
+      lockList.lockWriteAndCheckFullPath(lastInode, currentInodeDirectory, name);
```
do you have a guideline stating when we should lock and check parent, and when we should lock and check full path (or, parent + name) to replace original lock ?

calvinjia replied:
Its based on the caller's needs. If they want to ensure a particular inode is being accessed, then they should check the full path. If they are iterating over children, only the parent needs to be checked.

calvinjia commented on core/server/src/test/java/alluxio/master/file/meta/InodeFileTest.java:
```diff
@@ -167,4 +174,214 @@ public void permissionStatus() {
+    inode1.setName(name);
+    inode1.setParentId(dir1.getId());
+    inode1.lockReadAndCheckFullPath(dir1, name);
+    Assert.assertTrue(inode1.isReadLocked());
```
`lockReadAndCheckFullPath` only locks the particular inode.

calvinjia commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,27 +296,100 @@ public T setOwner(String owner) {
+   * @param parent the expected parent inode
+   * @throws InvalidPathException if the parent is not as expected
+   */
+  public void lockReadAndCheckParent(Inode parent) throws InvalidPathException {
```
Updated to have the same assumptions as lock read and check full path.

calvinjia commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,27 +296,100 @@ public T setOwner(String owner) {
+   * @param parent the expected parent inode
+   * @throws InvalidPathException if the parent is not as expected
+   */
+  public void lockWriteAndCheckParent(Inode parent) throws InvalidPathException {
```
Updated, thanks

calvinjia commented:
@apc999 Thanks for the review, I've address the comments.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12681/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12681/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12681/org.alluxio$alluxio-tests/testReport/alluxio.master.file/FileSystemMasterIntegrationTest/concurrentDelete/'><strong>alluxio.master.file.FileSystemMasterIntegrationTest.concurrentDelete</strong></a></li></ul>
Test FAILed.

calvinjia commented:
Jenkins, test this please.

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,27 +296,104 @@ public T setOwner(String owner) {
+   * consistent with what the caller is expecting. If the state is inconsistent, an exception
+   * will be thrown and the lock will be released.
+   *
+   * This method assumes that the inode path to the parent has been read locked.
```
nit: I would start this sentence with `NOTE: ...`

calvinjia replied:
Updated, thanks. x8

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/Inode.java:
```diff
@@ -277,27 +296,104 @@ public T setOwner(String owner) {
+   * consistent with what the caller is expecting. If the state is inconsistent, an exception
+   * will be thrown and the lock will be released.
+   *
+   * This method assumes that the inode path to the parent has been read locked.
```
ditto

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/InodeLockList.java:
```diff
@@ -41,6 +44,41 @@ public synchronized void lockRead(Inode<?> inode) {
+   * Locks the given inode in read mode, and adds it to this lock list. This method ensures the
+   * parent is the expected parent inode, and the name of the inode is the expected name.
+   *
+   * This method assumes that the inode path to the parent has been read locked.
```
`NOTE: ...`

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/InodeLockList.java:
```diff
@@ -68,6 +107,41 @@ public synchronized void lockWrite(Inode<?> inode) {
+   * Locks the given inode in write mode, and adds it to this lock list. This method ensures the
+   * parent is the expected parent inode.
+   *
+   * This method assumes that the inode path to the parent has been read locked.
```
`NOTE: ...`

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/InodeLockList.java:
```diff
@@ -68,6 +107,41 @@ public synchronized void lockWrite(Inode<?> inode) {
+   * Locks the given inode in write mode, and adds it to this lock list. This method ensures the
+   * parent is the expected parent inode, and the name of the inode is the expected name.
+   *
+   * This method assumes that the inode path to the parent has been read locked.
```
`NOTE: ...`

jsimsa commented on core/server/src/main/java/alluxio/master/file/meta/InodeTree.java:
```diff
@@ -695,9 +695,19 @@ private InodeLockList lockDescendantsInternal(InodeDirectory inodeDirectory,
+        try {
+          inodeGroup.lockWriteAndCheckParent(child, inodeDirectory);
+        } catch (InvalidPathException e) {
+          // Inode is no longer a descendant, continue.
```
ditto

calvinjia replied:
Updated, thanks.

jsimsa commented on core/server/src/test/java/alluxio/master/file/meta/InodeTreeTest.java:
```diff
@@ -217,9 +217,13 @@ public void createPathTest() throws Exception {
     // create nested directory
     InodeTree.CreatePathResult createResult =
-        createPath(mTree, NESTED_URI, sNestedDirectoryOptions);
+        createPath(mTree, NESTED_URI, dirOptions);
```
does this fit on the same line now?

calvinjia replied:
Updated, thanks.

jsimsa commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12684/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12686/
Test PASSed.

calvinjia commented on core/server/src/main/java/alluxio/master/file/meta/InodeTree.java:
```diff
@@ -695,9 +695,19 @@ private InodeLockList lockDescendantsInternal(InodeDirectory inodeDirectory,
+        try {
+          inodeGroup.lockReadAndCheckParent(child, inodeDirectory);
+        } catch (InvalidPathException e) {
+          // Inode is no longer a descendant, continue.
```
Updated, thanks.

calvinjia commented:
@jsimsa Thanks for the review, I've addressed the comments.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12688/
Test PASSed.

gpang commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1714,10 +1748,11 @@ public void rename(AlluxioURI srcPath, AlluxioURI dstPath) throws FileAlreadyExi
+    // modify operations on the parent inodes are thread safe.
+    try (InodePathPair inodePathPair =
+        mInodeTree.lockInodePathPair(srcPath, InodeTree.LockMode.WRITE, dstPath,
+            InodeTree.LockMode.WRITE)) {
```
Why is dst locked with WRITE? Don't we want the parent of dst to be READ locked? (And dst is not expected to already exist)

calvinjia replied:
You're right for cases when the source and destination parent are not the same, I've addressed this in #4466

gpang commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1830,65 +1865,116 @@ private long renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
+    srcInode.setName(srcName);
+    if (!srcParentInode.removeChild(srcInode)) {
+      // This should never happen.
+      LOG.error("Failed to rename within Alluxio. Alluxio and under storage may be inconsistent.");
```
The error message should include the src/dst path?

calvinjia replied:
Updated in #4466

gpang commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1830,65 +1865,116 @@ private long renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
+      LOG.error("Failed to rename within Alluxio. Alluxio and under storage may be inconsistent.");
+      srcInode.setName(dstName);
+      if (!dstParentInode.removeChild(dstName)) {
+        LOG.error("Failed to revert rename changes. Alluxio metadata may be inconsistent.");
```
The error message should include the src/dst path?

calvinjia replied:
Updated in #4466
